File: java/229-MajorityElementII.java
/*

Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times. The algorithm should run in linear time and in O(1) space.

*/

// http://www.geeksforgeeks.org/given-an-array-of-of-size-n-finds-all-the-elements-that-appear-more-than-nk-times/

class Solution{
  public List<Integer> majorityElement(int[] nums) {
        if (nums == null || nums.length == 0) return new ArrayList<Integer>();
        List<Integer> result = new ArrayList<Integer>();
        int number1 = nums[0], number2 = nums[0], count1 = 0, count2 = 0, len = nums.length;
        for (int i = 0; i < len; i++) {
            if (nums[i] == number1)
                count1++;
            else if (nums[i] == number2)
                count2++;
            else if (count1 == 0) {
                number1 = nums[i];
                count1 = 1;
            } else if (count2 == 0) {
                number2 = nums[i];
                count2 = 1;
            } else {
                count1--;
                count2--;
            }
        }
        count1 = 0;
        count2 = 0;
        for (int i = 0; i < len; i++) {
            if (nums[i] == number1)
                count1++;
            else if (nums[i] == number2)
                count2++;
        }
        if (count1 > len / 3)
            result.add(number1);
        if (count2 > len / 3)
            result.add(number2);
        return result;
    }    
}


class Solution1 {
    public List<Integer> majorityElement(int[] nums) {
        int n = nums.length;
        int res1 = 0, res2 = 0;
        int c1 = -1, c2 = -1;
        for(int i = 0; i<n; i++) {
            if(c1 == -1) {
                c1 = 1;
                res1 = nums[i];
            }
            else if(nums[i] == res1) {++c1;}
            else if(c2 == -1) {
                c2 = 1;
                res2 = nums[i];
            }
            else if(nums[i] == res2) {
                ++c2;
            }
            else if (c1 == 0) {
                res1 = nums[i];
                ++c1;
            }
            else if (c2 == 0) {
                res2 = nums[i];
                ++c2;
            }
            else {
                --c1;
                --c2;
            }
        }
        c1 = 0; c2 = 0;
        for(int i = 0; i<n; i++) {
            if(nums[i] == res1) ++c1;
            if(nums[i] == res2) ++c2;
        }
        List<Integer> res = new ArrayList<Integer>();
        if(c1>n/3) res.add(res1);
        if(c2>n/3 && res2!=res1) res.add(res2);
        return res;
    }
}

File: c++/229-MajorityElementII.cpp
/*

Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times. The algorithm should run in linear time and in O(1) space.

*/

// http://www.geeksforgeeks.org/given-an-array-of-of-size-n-finds-all-the-elements-that-appear-more-than-nk-times/
class Solution1 {
public:
    vector<int> majorityElement(vector<int>& nums) {
    int cnt1 = 0, cnt2 = 0, a=0, b=1;
    
    for(auto n: nums){
        if (a==n){
            cnt1++;
        }
        else if (b==n){
            cnt2++;
        }
        else if (cnt1==0){
            a = n;
            cnt1 = 1;
        }
        else if (cnt2 == 0){
            b = n;
            cnt2 = 1;
        }
        else{
            cnt1--;
            cnt2--;
        }
    }
    
    cnt1 = cnt2 = 0;
    for(auto n: nums){
        if (n==a)   cnt1++;
        else if (n==b)  cnt2++;
    }
    
    vector<int> res;
    if (cnt1 > nums.size()/3)   res.push_back(a);
    if (cnt2 > nums.size()/3)   res.push_back(b);
    return res;
}
};

class Solution2 {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int k = 3;
        const int n = nums.size();
        unordered_map<int, int> hash;

        for (const auto& i : nums) {
            ++hash[i];
            if (hash.size() == k) {
                auto it = hash.begin();
                while (it != hash.end()) {
                    if (--(it->second) == 0) {
                        hash.erase(it++);
                    } else {
                        ++it;
                    }
                }
            }
        }

        // Resets hash for the following counting.
        for (auto& it : hash) {
            it.second = 0;
        }

        // Counts the occurrence of each candidate integer.
        for (const auto& i : nums) {
            auto it = hash.find(i);
            if (it != hash.end()) {
                ++it->second;
            }
        }

        // Selects the integer which occurs > [n / k] times.
        vector<int> ret;
        for (const pair<int, int>& it : hash) {
            if (it.second > n / k) {
                ret.push_back(it.first);
                //ret.emplace_back(it.first);
            }
        }
        return ret;
    }
};

class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int k = 3;
        const int n = nums.size();
        unordered_map<int, int> hash;

        for (const auto& i : nums) {
            ++hash[i];
            // if (hash.size() == k+1) { 
            if (hash.size() == k) { // both k and k+1 work; just increasing candidate size   
                for (auto it = hash.begin(); it != hash.end(); ) {
                    if (--(it->second) == 0) {
                        it = hash.erase(it); //works bcz erase returns next iterator
                        /*
                        This doesn't work : hash.erase(it); ++it;
                        bcz of itertaor invalidation
                        see http://stackoverflow.com/questions/263945/what-happens-if-you-call-erase-on-a-map-element-while-iterating-from-begin-to
                        This works: hash.erase(it++);
                        bcz it++ increments it and returns previous it value to erase. 
                        */
                    } 
                    else {
                        ++it;
                    }
                }
            }            
        }
        
        // Resets hash for the following counting.
        for (auto& it : hash) {
            it.second = 0;
        }

        // Counts the occurrence of each candidate integer.
        for (const auto& i : nums) {
            auto it = hash.find(i);
            if (it != hash.end()) {
                ++it->second;
            }
        }

        // Selects the integer which occurs > [n / k] times.
        vector<int> ret;
        for (const pair<int, int>& it : hash) {
            if (it.second > n / k) {
                ret.push_back(it.first);
                //ret.emplace_back(it.first);
            }
        }
        return ret;
    }
};
